cola/widgets/branch.py

```diff
@@ -0,0 +1,520 @@
+                item = BranchTreeWidgetItem(key, child_icon)
+                item.addChildren(create_children(grouped_branches[key]))
+
+                if item.childCount() > 0:
+                    item.setIcon(0, icons.ellipsis())
+
+                result.append(item)
+
+            result.sort(key=lambda x: x)
+
+            return result
+
+        branch = BranchTreeWidgetItem(name, icons.ellipsis())
+        branch.addChildren(create_children(dict_children))
+
+        return branch
+
+    @staticmethod
+    def get_root(item):
+        """Returns top level item from an item"""
+        parents = [item]
+        parent = item.parent()
+
+        while parent is not None:
+            parents.append(parent)
+            parent = parent.parent()
+
+        return parents[len(parents) - 1]
+
+    @staticmethod
+    def get_full_name(item, join_char):
+        """Returns item full name generated by iterating over
+        parents and joining their names with 'join_char'"""
+        parents = [item.name]
+        parent = item.parent()
+
+        while parent is not None:
+            parents.append(parent.name)
+            parent = parent.parent()
+
+        result = join_char.join(reversed(parents))
+
+        return result[result.find(join_char) + 1:]
+
+    @staticmethod
+    def expand_from_item(item):
+        """Expand tree parents from item"""
+        parent = item.parent()
+
+        while parent is not None:
+            parent.setExpanded(True)
+            parent = parent.parent()
+
+    def load_state(self, item, state):
+        """Load expanded items from a dict"""
+        if len(state.keys()) > 0:
+            item.setExpanded(True)
+
+        for i in range(item.childCount()):
+            child = item.child(i)
+            if child.name in state:
+                self.load_state(child, state[child.name])
+
+    def save_state(self, item):
+        """Save expanded items in a dict"""
+        result = {item.name: {}}
+
+        if item.isExpanded():
+            for i in range(item.childCount()):
+                child = item.child(i)
+                result[item.name].update(self.save_state(child))
+
+        return result
+
+
+class GitHelper(object):
+    def __init__(self, git):
+        self.git = git
+
+    def log(self, origin, args):
+        return self.git.log(origin, *args)
+
+    def fetch(self, remote, branch):
+        return self.git.fetch(remote, branch, verbose=True)
+
+    def push(self, remote, branch):
+        return self.git.push(remote, branch, verbose=True)
+
+    def pull(self, remote, branch):
+        return self.git.pull(remote, branch, no_ff=True, verbose=True)
+
+    def delete_remote(self, remote, branch):
+        return self.git.push(remote, branch, delete=True)
+
+    def delete_local(self, branch):
+        return self.git.branch(branch, D=True)
+
+    def merge(self, branch):
+        return self.git.merge(branch, no_ff=False, no_commit=True)
```

Comment:
If you say `no_ff=False` then it effectively does nothing -- no flag is every passed to git.  If you want the `--no-ff` flag to make it through, set it to `True`.